timely_beliefs/sensors/func_store/knowledge_horizons.py

```diff
@@ -31,6 +34,45 @@ def at_date(
     return event_start - knowledge_time.astimezone(event_start.tzinfo)
 
 
+def x_years_ago_at_date(
+    event_start: datetime | pd.DatetimeIndex,
+    day: int,
+    month: int,
+    x: int = 1,
+    get_bounds: bool = False,
+    z: str | None = None,
+) -> timedelta | pd.TimedeltaIndex | tuple[timedelta, timedelta]:
+    """Compute the sensor's knowledge horizon to represent the event could be known since some date, `x` years ago.
+
+    For example, it can be used for a tax rate that changes annually and with a known publication date.
+
+    :param event_start:     Start of the event, used as an anchor for determining the knowledge horizon.
+    :param day:             Reference day of the month of the annual date to compare against.
+    :param month:           The month of the annual date to compare against.
+    :param x:               The number of years to shift the reference date to.
+    :param z:           Timezone string.
+    :param get_bounds:      If True, this function returns bounds on the possible return value.
+                            These bounds are normally useful for creating more efficient database queries when filtering by belief time.
+    """
+
+    MAX_DAYS_IN_A_YEAR = 366
+    MIN_DAYS_IN_A_YEAR = 364
+
+    if x <= 0:
+        raise ValueError("Only positive values for `x` are supported.")
```

Comment:
After the following thought exercise, I believe we almost support zero and negative x out of the box.

For x=0:

- The last possible knowledge time is 31st of December, which, when combined with the first possible event in the same year, starting on January 1st, should yield a minimum knowledge horizon of -365 days (i.e. knowledge time 365 days after the event started), for a leap year.
- Likewise, the first possible knowledge time is 1st of January, which, when combined with the last possible event in the same year, starting just before new year's midnight, should yield a maximum knowledge horizon of nearly 366 days (i.e. knowledge time 366 days before the event starts), for a leap year.

For x=-1, which signifies the event could be known no sooner than some date in the year following the event start, we have:

- Event starts just before new year's midnight and knowledge time on January 1st: bound is 0 days.
- Event starts January 1st and knowledge time on December 31st the following year: bound is minus 2 full years.

Reply:
I'll follow up later with an extension, but for now we can keep the NotImplementedError.